### PR TITLE
[Video] The controls bar flickers when video's popover menu is closed

### PR DESCRIPTION
**Pull request by ikevin127**
### Details

Because within our [VideoPopoverMenu](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/VideoPlayer/BaseVideoPlayer.tsx#L326-L330) component we're using a [portal](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/Popover/index.tsx#L58-L75), when this is **open** it comes on top of the `BaseVideoPlayer` Hoverable surface of the video. These are the events that occur which lead to our issue:

- The video player controls 3-dot menu (More) is clicked -> Popover opens -> isHovered state is set to false.
- User clicks outside of the Popover -> Popover closes -> isHovered state is set to false again.
- Right after the Popover is gone, the Hoverable surface re-captures the mouse event -> isHovered state is set to true.

Our issue occurs between (2.) and (3.) in very quick succession which causes the flicker of the video player controls bar because of this condition [here](https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/components/VideoPlayer/BaseVideoPlayer.tsx#L308), specifically the isHovered state variable coming from the Hoverable component which wraps the video container, toggling false/true in quick succession right after the isPopoverVisible becomes false.

This PR introduces a new boolean prop to the `Hoverable` component, named `shouldFreezeCapture` which will be used within the `ActiveHoverable` component to prevent setting the isHovered state to false the second time (unnecessarily) when an open portal (popover / modal) is being closed, essentially freezing the capture of the hover event.

> [!note]
> This behaviour and fix can observed only on large layout devices where hover event is available.

### Fixed Issues

$ https://github.com/Expensify/App/issues/39598
PROPOSAL: https://github.com/Expensify/App/issues/39598#issuecomment-2041187706

### Tests

> [!tip]
> 1. Go to any chat.
> 2. Add a video attachment (you don't need to upload it).
> 3. Click on the 3-dot (`More`) button to open the Popover (you should see the `Playback speed` option).
> 4. Click outside of the Popover in order to close it.
> 5. Notice:
> - if the click outside the popover was applied **within** the video view -> the video controls bar remains **visible** and does not flicker when the popover menu is closed;
> - if the click outside the popover was applied **outside** of the video view -> the video controls bar **disappears** without flickering

- [x] Verify that no errors appear in the JS console

### Offline tests

**TLDR:** same as **Tests**.

> [!tip]
> 1. Go to any chat.
> 2. Add a video attachment (you don't need to upload it).
> 3. Click on the 3-dot (`More`) button to open the Popover (you should see the `Playback speed` option).
> 4. Click outside of the Popover in order to close it.
> 5. Notice:
> - if the click outside the popover was applied **within** the video view -> the video controls bar remains **visible** and does not flicker when the popover menu is closed;
> - if the click outside the popover was applied **outside** of the video view -> the video controls bar **disappears** without flickering

### QA Steps

**TLDR:** same as **Tests**.

> [!tip]
> 1. Go to any chat.
> 2. Add a video attachment (you don't need to upload it).
> 3. Click on the 3-dot (`More`) button to open the Popover (you should see the `Playback speed` option).
> 4. Click outside of the Popover in order to close it.
> 5. Notice:
> - if the click outside the popover was applied **within** the video view -> the video controls bar remains **visible** and does not flicker when the popover menu is closed;
> - if the click outside the popover was applied **outside** of the video view -> the video controls bar **disappears** without flickering

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/56457735/839d5338-8c91-4d47-b511-d1c556e73af1

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/56457735/1301d581-1ea4-4ed7-a87b-684afb7afe2e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/95691ba4-1a68-4328-b26c-eabb0dfee80f

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/56457735/8093d22a-f30e-4b9a-af23-a9b868ae1d1d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/52792167-991c-47b6-8334-df47971b6ccb

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/56457735/2a47ddbb-db73-44e9-80f9-2111e75ff25e

</details>

### CODE REVIEW

**Review by rojiphil — COMMENTED** [review]
@ikevin127 Observations on testing the PR version with expanded mode of video player are as below:

**1) Chrome:** 
a) Click inside of player => Action Bar remains shown (without flicker)
b) Click outside of player => Action Bar hides
<details>
<summary>Chrome:Outside Click</summary>

https://github.com/Expensify/App/assets/3004659/237305f0-26fc-49c1-bb48-be75dbd45780

</details>

**2) Safari:**
a) Click inside of player => Action Bar remains shown (without flicker)
b) Click outside of player => Action Bar remains shown. Only after mouse move, it hides. 
<details>
<summary>Safari:Outside Click</summary>

https://github.com/Expensify/App/assets/3004659/242faa2b-e5b0-4269-8738-27f398c4392b

</details>

So, while the flicker issue is resolved, we still have an inconsistency issue between Chrome and Safari when clicked outside. Here, the expected behavior is for the Action bar to be hidden on clicking outside on Safari (i.e. just like Chrome does).
What do you think?

**Comment by ikevin127** [comment]
@rojiphil I think given that the flicker issue is resolved, the behaviour on Safari is strictly related to browser-specific handling of the `mouseover` event and not a blocker when it comes to this PR.

From what I can see, on Safari, once the popover is closed, the user needs to actually move the cursor outside of the video view for the controls to dissapear which means that the Hoverable component only captures the mouse moving once the user actually moves it.

I think these are just small differences between how Safari vs other browsers handle the [`mouseover`](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseover_event) event capture.

If you think this is a PR blocker then let me know and I'll spend a few more days looking into this in hopes to align the behaviour from Safari with other browsers.

**Comment by rojiphil** [comment]
> the behaviour on Safari is strictly related to browser-specific handling of the mouseover event and not a blocker when it comes to this PR

@ikevin127 The behavior on Safari changed only after this PR. Before the fix, the action bar hid when clicked outside the player. But now it doesn't. To me, this is a regression and needs a fix. no?

**Comment by ikevin127** [comment]
@rojiphil Oh, I see - I did not know that it changed because of this PR. Will definitely look into it and fix it in this case!

**Comment by ikevin127** [comment]
@rojiphil I looked into it and I can't find a solution. The problem is that Safari doesn't re-capture the `mouseover` event when the popover is closed without user actually moving the cursor, whereas Chrome re-captures it without user action. 
Here's the behaviour on Safari before the fix (current main) and with the fix:

<details>
<summary>MacOS: Safari (before / after)</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/25cdb613-6a71-48bf-a56a-74a4332c14d1"> | <video src="https://github.com/Expensify/App/assets/56457735/b4fe3b0f-5d11-4a0e-845f-ce1c5d5e8c5a"> |

</details>

As you can see on the before video on Safari, the issue on current main is that when the popover closes the video controls are hidden and only show up when the user moves the mouse over the video. With the fix, this issue is solved, but this thing will happen now with hiding the controls when cursor is outside -> which only works when user moves the mouse outside of the video in order for Safari to re-capture the `mousemove` event and hide the video controls.

Again, this is a weird Safari specific issue where cursor-related events are not fired unless the user actually moves the cursor. To conclude, it's either you're fine with moving forward with this as is, or I'm going to close this PR and the issue can continue to look for other proposals which will cover everything you're asking for, perfectly.

I, for one, I'm not willing to invest anymore more time into this as my capacity is limited and additionally -> I consider the issue is fixed and the behaviour on Safari evens out if we compare before / after.

Let me know what you decide!

**Comment by rojiphil** [comment]
> it's either you're fine with moving forward with this as is

The worry I have here is that the PR changes have changed behavior as mentioned [here](https://github.com/Expensify/App/pull/41125#pullrequestreview-2026700340) and [here](https://github.com/Expensify/App/pull/41125#issuecomment-2081847639). However, since it is only until the mouse moves again, it may also be fine. I would want @mountiny to weigh in here and arrive at a consensus. What’s your take here?

**Comment by rojiphil** [comment]
Meanwhile, let me also get a quick check on this with @Skalakid. Do you see any way to fix the behavior mentioned [here](https://github.com/Expensify/App/pull/41125#pullrequestreview-2026700340)?

**Comment by Skalakid** [comment]
I agree with @ikevin127 it's a really small edge case and it seems like a Safari bug. I tried using the `mouseover` event dispatching and updating the `isHovered` state when `shouldFreezeCapture` is set back to false, but every time there was a problem with event order on different browsers and new edge cases appeared. I can look into that tomorrow

**Comment by Skalakid** [comment]
I wasn't able to find any solution for [this problem](https://github.com/Expensify/App/pull/41125#pullrequestreview-2026700340) in my limited time capacity. @mountiny should we go with current solution or try to get different proposal that doesn't introduce any regression?

**Comment by ikevin127** [comment]
> try to get different proposal that doesn't introduce any regression

I wouldn't call the observed behaviour a `regression` 😅
I see the current vs PR behaviour on Safari as an actual `improvement`, as detailed in https://github.com/Expensify/App/pull/41125#issuecomment-2083778317 (see before / after videos):

> As you can see on the before video on Safari, the issue on current main is that when the popover closes the video controls are hidden and only show up when the user moves the mouse over the video. With the fix, this issue is solved, but this thing will happen now with hiding the controls when cursor is outside -> which only works when user moves the mouse outside of the video in order for Safari to re-capture the mousemove event and hide the video controls.

**Comment by mountiny** [comment]
@rojiphil @ikevin127 I dont think [this](https://github.com/Expensify/App/pull/41125#pullrequestreview-2026700340) is a problem, we can go ahead with this PR if that is the only thing noticed

**Comment by rojiphil** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/3004659/29d901e6-b57a-47b1-9ab7-33a4f84c86ad

https://github.com/Expensify/App/assets/3004659/7e23e8f5-790b-4fa6-a10d-908a2cde5333

</details>

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/3004659/cb93e9e8-7309-4bdb-960e-4f2f60408d16

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/3004659/41a841dc-8d59-4b36-bf5d-74a49c9c81bc

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/3004659/fd22d7a3-b29c-4a74-a478-7b13f73a630d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/3004659/2d623e2d-c601-4201-ae77-9a81d03659b4

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/3004659/c3a959f7-dae8-44ac-b5e5-a805f313e747

</details>

**Review by rojiphil — APPROVED** [review]
Thanks @ikevin127. LGTM and tests well too.

**Review by mountiny — APPROVED** [review]
Thanks!
